SocketTest fails with unexpected exception

This is probably an issue with my box with a port taken or something but the test fails in perhaps an alternative path. I am not sure if this is a failure or not.

Config:
macOS Ventura 13.7
openjdk 17.0.12 2024-07-16 LTS
Apple clang version 14.0.3 (clang-1403.0.22.14.1)


```scala
[error] Test org.scalanative.testsuite.javalib.net.SocketTest.connectWithTimeout failed: java.lang.AssertionError: unexpected exception type thrown; expected:<java.net.SocketTimeoutException> but was:<java.net.ConnectException>, took 0.008 sec
[error]     at java.lang.AssertionError.this(Throwables.scala:209)
[error]     ...
[error] Caused by: java.net.ConnectException: Connection refused
```

I am not sure but I usually get this dialog twice but if I click it away it used to work. Even after several tries I have no luck. I will try tonight on a different network.

![Image](https://github.com/user-attachments/assets/bc2e5b24-f713-43ec-bc55-dc1622dd5037)

I did try and change that setting as suggested.

![Image](https://github.com/user-attachments/assets/e2d84d0f-5966-47d8-b186-8865a29d1dd2)

I clicked way but still get the issue.

```scala
[error] Test org.scalanative.testsuite.javalib.net.SocketTest.connectWithTimeout failed: java.lang.AssertionError: unexpected exception type thrown; expected:<java.net.SocketTimeoutException> but was:<java.net.ConnectException>, took 0.008 sec
[error]     at java.lang.AssertionError.this(Throwables.scala:209)
[error]     ...
[error] Caused by: java.net.ConnectException: Connection refused
[error]     at java.net.UnixPlainSocketImpl.loop$1(UnixPlainSocketImpl.scala:52)
[error]     at java.net.UnixPlainSocketImpl.tryPollOnConnect(UnixPlainSocketImpl.scala:38)
[error]     at java.net.AbstractPlainSocketImpl.connect6(AbstractPlainSocketImpl.scala:217)
[error]     at java.net.AbstractPlainSocketImpl.$anonfun$connectFunc$2(AbstractPlainSocketImpl.scala:262)
[error]     at java.net.AbstractPlainSocketImpl$$Lambda$9.apply(AbstractPlainSocketImpl.scala:262)
[error]     at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.scala:274)
[error]     at java.net.Socket.startup(Socket.scala:102)
[error]     at java.net.Socket.connect(Socket.scala:153)
[error]     at org.scalanative.testsuite.javalib.net.SocketTest.$anonfun$connectWithTimeout$1(SocketTest.scala:231)
[error]     at org.scalanative.testsuite.javalib.net.SocketTest$$Lambda$2.apply(SocketTest.scala:231)
[error]     at org.scalanative.testsuite.utils.AssertThrows$$anon$1.run(AssertThrows.scala:29)
[error]     ... 49 more
```

```scala
[error] Test org.scalanative.testsuite.javalib.net.DatagramSocketTest.sendReceiveBroadcast failed: java.net.SocketTimeoutException: receive timed out, SO_TIMEOUT: 500, took 0.509 sec
[error]     at java.net.UnixPlainDatagramSocketImpl.tryPoll(UnixPlainDatagramSocketImpl.scala:45)
[error]     at java.net.AbstractPlainDatagramSocketImpl.recvfrom(AbstractPlainDatagramSocketImpl.scala:278)
[error]     at java.net.AbstractPlainDatagramSocketImpl.receive(AbstractPlainDatagramSocketImpl.scala:326)
[error]     at java.net.DatagramSocket.receive(DatagramSocket.scala:216)
[error]     at org.scalanative.testsuite.javalib.net.DatagramSocketTest.$anonfun$sendReceiveBroadcast$4(DatagramSocketTest.scala:410)
[error]     at org.scalanative.testsuite.javalib.net.DatagramSocketTest$$Lambda$11.apply(DatagramSocketTest.scala:410)
[error]     at scala.Option.foreach(Option.scala:437)
[error]     at org.scalanative.testsuite.javalib.net.DatagramSocketTest.sendReceiveBroadcast(DatagramSocketTest.scala:398)
[error]     at org.scalanative.testsuite.javalib.net.DatagramSocketTest$scalanative$junit$bootstrapper$.invokeTest(DatagramSocketTest.scala:26)
[error]     ... 44 more
```

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/DatagramSocketTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/DatagramSocketTest.scala
@@ -430,10 +430,38 @@ class DatagramSocketTest {
 
           val result =
             new DatagramPacket(Array.ofDim[Byte](bytes.length), bytes.length)
-          ds2.receive(result)
+          try {
+            ds2.receive(result)
 
-          val receivedData = new String(result.getData())
-          assertEquals("Received incorrect data", data, receivedData)
+            val receivedData = new String(result.getData())
+            assertEquals("Received incorrect data", data, receivedData)
+          } catch {
+            /* DatagramSockets use UDP transport, which is "send-and-pray".
+             *
+             * This Test is designed for the Scala Native Continuous
+             * Integration environment, where it passes.
+             *
+             * In other environments a timeout may be due to a number
+             * of causes. Some, not all, possibilities. It might be:
+             *
+             *   - Today is not the day for answered prayers.
+             *
+             *   - Today is just a slow day, packets would have been received
+             *     if the timeout was longer.
+             *
+             *   - This Test is validly detecting lost packets, say send
+             *     and receive addresses not matching up.
+             *
+             *   - A defect in this Test, attempting to send or receive
+             *     on an interface where wither or both are disabled.
+             *
+             *   - A firewall or other intermediate is routing the
+             *     sent packages to a black hole sink, so they never
+             *     arrive.
+             */
+            case ex: SocketTimeoutException =>
+              fail("receive timed out, check network & firewall configuration")
+          }
         } finally {
           ds1.close()
           ds2.close()

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/DatagramSocketTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/DatagramSocketTest.scala
@@ -395,6 +395,26 @@ class DatagramSocketTest {
   }
 
   @Test def sendReceiveBroadcast(): Unit = {
+    /* Issue 4221
+     *   This Test should be run only in environments where:
+     *   1) It is responsible to broadcast packets to every node on net.
+     *   2) The network configuration, possible firewalls, and routers
+     *      allow broadcasting.
+     *
+     *   This is true in the Scala Native Continuous Integration environment
+     *   but may not be in more general work environments.
+     *
+     *   Network developers will need to bypass this check and run the
+     *   test manually.
+     */
+    assumeTrue(
+      "Advanced and/or CI-only test",
+      ju.Optional
+        .ofNullable(System.getenv("GITHUB_ACTIONS"))
+        .orElse("false")
+        .equalsIgnoreCase("true")
+    )
+
     // NetworkInterface.getNetworkInterfaces is not implemented in Windows
     assumeFalse("Not implemented in Windows", Platform.isWindows)
     assumeNotRoot()

--- a/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/SocketTest.scala
+++ b/unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/net/SocketTest.scala
@@ -227,7 +227,8 @@ class SocketTest {
     try {
       assertThrows(
         classOf[SocketTimeoutException],
-        s.connect(new InetSocketAddress("123.123.123.123", 12341), 100)
+        // Use a document-only Internet address and very short timeout.
+        s.connect(new InetSocketAddress("203.0.113.1", 12341), 2)
       )
     } finally {
       s.close()

